### PR TITLE
Quincy: bluestore & bluestore-rdr to Elastic Shared Blobs

### PR DESCRIPTION
**Pull request by aclamk**
This is continuation of work in:
https://github.com/rzarzynski/ceph/commits/wip-os-bluestore-exp-quincy2
which in turn bases on 
https://github.com/ceph/ceph/pull/51451

New coding to allow tools to work on both bluestore and bluestore-rdr.
Related tools are: ceph-kvstore-tool and ceph-bluestore-tool.

Minor fixes to rdr implementation.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by aclamk** [comment]
Teuthology runs:
Original:
https://pulpito.ceph.com/akupczyk-2023-06-13_12:20:40-rados-wip-aclamk-os-bluestore-rdr-quincy-distro-default-smithi/
Branch which always creates "bluestore-rdr":
https://pulpito.ceph.com/akupczyk-2023-06-14_12:36:55-rados-wip-aclamk-os-bluestore-rdr-quincy-on-distro-default-smithi/

Failures analyzed in (huge thanks to Laura):
https://pad.ceph.com/p/akupczyk-teuthology-analysis

**Comment by rzarzynski** [comment]
```
FAILED: src/os/CMakeFiles/os_exp.dir/bluestore/BlueStore.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DHAVE_CONFIG_H -DWITH_EXPERIMENTAL -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -isystem /opt/ceph/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/rapidjson/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/rocksdb/include -isystem /usr/include/fuse -Werror -g -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++17 -MD -MT src/os/CMakeFiles/os_exp.dir/bluestore/BlueStore.cc.o -MF src/os/CMakeFiles/os_exp.dir/bluestore/BlueStore.cc.o.d -o src/os/CMakeFiles/os_exp.dir/bluestore/BlueStore.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/BlueStore.cc
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/BlueStore.cc:6956:14: error: 'ceph::experimental::ZonedAllocator' is an incomplete type
    auto a = dynamic_cast<ZonedAllocator*>(alloc);
```

https://jenkins.ceph.com/job/ceph-pull-requests/116940/consoleFull#13632025982a811ea2-3e7b-466b-84b4-d13df7e35809

File: src/os/bluestore/bluestore_tool.cc [hunk 1]
```diff
@@ -72,6 +73,53 @@ void validate_path(CephContext *cct, const string& path, bool bluefs)
   }
 }
 
+struct BlueStorePair
```
Comment (rzarzynski) [inline src/os/bluestore/bluestore_tool.cc:76]: There is the `static_ptr<...>` template that does exactly this without the need for dynamic allocation.
On the other hand, the current approach is so _explicitly hackish_ I like it while nobody really cares dynallocs in the tools.

File: src/os/bluestore/bluestore_tool.cc [hunk 2]
```diff
@@ -681,7 +681,8 @@ int main(int argc, char **argv)
     // kludge some things into the map that we want to populate into
     // target dir
     label.meta["path_block"] = devs.front();
-    label.meta["type"] = "bluestore";
+    //now type is stored in label, this allows bluestore-rdr
+    //label.meta["type"] = "bluestore";
```
Comment (rzarzynski) [inline src/os/bluestore/bluestore_tool.cc:685]: OK, so now we're leaving the key intact. Is there any business in  checking whether it's `buestore` or `bluestore-rdr`?

File: src/os/bluestore/bluestore_tool.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/bluestore_tool.cc:76]: What good does come from using `static_ptr` ?
The problem is that bluestore and bluestore-rdr are 2 completely independent classes.
Granted, I can keep it in one place; but then I would need to `reinterpret_cast` the location.

I agree that the current solution is _crude and primitive_ but not _hackish_.

File: src/os/bluestore/bluestore_tool.cc [hunk 3]
```diff
@@ -72,6 +73,53 @@ void validate_path(CephContext *cct, const string& path, bool bluefs)
...
+  }
+  ceph_assert(bsp.b1 == nullptr || bsp.b2 == nullptr);
+  return bsp;
+}
+
+// attempt action
+#define BLUESTORE_ACTION(bsp, action) { if (bsp.b1) { bsp.b1->action; } else { bsp.b2->action; } }
+
+// attempt func
+#define BLUESTORE_FUNC(bsp, func) (bsp.b1 ? bsp.b1->func : bsp.b2->func)
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/bluestore_tool.cc:121]: I do not know how. The only way I know is a non-standard gcc extension `({int x;....;x;})`

File: src/os/bluestore/bluestore_tool.cc (hunk 2 above)
Comment (aclamk) [reply src/os/bluestore/bluestore_tool.cc:685]: There is - we do not want an accidental upgrade from pacific to quincy-rdr.

**Comment by aclamk** [comment]
jenkins test make check

**Comment by aclamk** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.h [hunk 4]
```diff
@@ -2843,8 +2845,8 @@ class BlueStore : public ObjectStore,
   void _zoned_cleaner_stop();
   void _zoned_cleaner_thread();
   void _zoned_clean_zone(uint64_t zone_num,
-			 class ZonedAllocator *a,
-			 class ZonedFreelistManager *f);
+			 ZonedAllocator *a,
+			 ZonedFreelistManager *f);
```
Comment (rzarzynski) [inline src/os/bluestore/BlueStore.h:2849]: ACK.

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -57,12 +57,22 @@
 #include "ZonedFreelistManager.h"
 #endif
 
-#if defined(WITH_LTTNG) && !defined(WITH_EXPERIMENTAL)
+#if defined(WITH_LTTNG)
```
Comment (rzarzynski) [inline src/os/bluestore/BlueStore.cc:60]: ACK, THX!

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -8269,7 +8269,7 @@ int BlueStore::mkfs()
     r = read_meta("type", &type);
     if (r == 0) {
       if (type != BS_FLAVOR) {
-	derr << __func__ << " expected bluestore, but type is " << type << dendl;
+	derr << __func__ << " expected " << BS_FLAVOR << ", but type is " << type << dendl;
```
Comment (rzarzynski) [inline src/os/bluestore/BlueStore.cc:8272]: ACK.

**Review by rzarzynski — APPROVED** [review]
The recent `bluestore-rdr` changes LGTM!

**Comment by aclamk** [comment]
Successful teuthology runs:
https://pulpito.ceph.com/akupczyk-2023-06-14_12:36:55-rados-wip-aclamk-os-bluestore-rdr-quincy-on-distro-default-smithi/
and
https://pulpito.ceph.com/akupczyk-2023-06-13_12:20:40-rados-wip-aclamk-os-bluestore-rdr-quincy-distro-default-smithi/

Analysis in: https://pad.ceph.com/p/akupczyk-teuthology-analysis

The only really failed test is an upgrade from pacific to bluestore-rdr. It should have failed, because new bluestore was forced to be bluestore-rdr. The only error was an improper error message:
https://pulpito.ceph.com/akupczyk-2023-06-14_12:36:55-rados-wip-aclamk-os-bluestore-rdr-quincy-on-distro-default-smithi/7304206

**Review by guits — APPROVED** [review]
ceph-volume changes approved

https://pulpito.ceph.com/gabrioux-2023-07-11_06:48:53-orch:cephadm-cv-bluestore-rdr-quincy-distro-default-smithi/
https://pulpito.ceph.com/gabrioux-2023-07-11_12:54:10-orch:cephadm-cv-bluestore-rdr-quincy-distro-default-smithi/

**Comment by BlaineEXE** [comment]
For my and other Rook users' reference, the image built by Ceph CI is made available here:
 
https://quay.ceph.io/repository/ceph-ci/ceph?tab=tags&tag=wip-aclamk-os-bluestore-rdr-quincy

**Pull request closed without merging**